protocol/client/stream.go

```diff
@@ -43,114 +32,150 @@ func (pub PublicStream) Handle(msg protocol.Message) error {
 	return nil
 }
 
-func NewPrivateHandler(db Database) StreamHandler {
-	return PrivateStream{
-		db: db,
-	}.Handle
-}
-
-// PrivateStream is a stream with messages from multiple sources.
-// In our case every message will have a pubkey (derived from signature) that will be used
-// to determine who is the writer
-type PrivateStream struct {
-	db Database
+// StreamHandlerMultiplexed handles a stream of messages
+// where a sender of a message (Contact) is figured out
+// from the message itself based on a public key used
+// to sign a message.
+func StreamHandlerMultiplexed(db Database) StreamHandler {
+	return func(m *protocol.Message) error {
+		return streamHandlerMultiplexed(db, m)
+	}
 }
 
-func (priv PrivateStream) Handle(msg protocol.Message) error {
-	if msg.SigPubKey == nil {
-		return errors.New("message should be signed")
+func streamHandlerMultiplexed(db Database, m *protocol.Message) error {
+	publicKey := m.SigPubKey
+	if publicKey == nil {
+		return errors.New("message is unsigned")
 	}
-	keyhex := PubkeyToHex(msg.SigPubKey)
+
 	contact := Contact{
 		Type:      ContactPublicKey,
 		State:     ContactNew,
-		Name:      keyhex, // TODO(dshulyak) replace with 3-word funny name
-		PublicKey: msg.SigPubKey,
+		Name:      pubkeyToHex(publicKey), // TODO(dshulyak) replace with 3-word funny name
+		PublicKey: publicKey,
 		Topic:     DefaultPrivateTopic(),
 	}
-	exist, err := priv.db.PublicContactExist(contact)
+
+	exists, err := db.PublicContactExist(contact)
 	if err != nil {
-		return errors.Wrap(err, "error verifying if public contact exist")
+		return errors.Wrap(err, "error verifying if a contact exist")
 	}
-	if !exist {
-		err := priv.db.SaveContacts([]Contact{contact})
+
+	if !exists {
+		err := db.SaveContacts([]Contact{contact})
+		if err != nil {
+			return errors.Wrap(err, "can't save a new contact")
+		}
+	} else {
+		// TODO: replace with db.ContactByPublicKey()
+		contacts, err := db.Contacts()
 		if err != nil {
-			return errors.Wrap(err, "can't add new contact")
+			return errors.Wrap(err, "error getting contacts")
+		}
+		for _, c := range contacts {
+			if c.PublicKey == nil {
+				continue
+			}
+
+			// TODO: extract
+			if publicKey.X.Cmp(c.PublicKey.X) == 0 && publicKey.Y.Cmp(c.PublicKey.Y) == 0 {
+				contact = c
+				break
+			}
 		}
 	}
-	// TODO discard message from blocked contact (State != ContactBlocked)
-	_, err = priv.db.SaveMessages(contact, []*protocol.Message{&msg})
+
+	// TODO: discard message from blocked contact (State != ContactBlocked)
+	_, err = db.SaveMessages(contact, []*protocol.Message{m})
 	if err == ErrMsgAlreadyExist {
 		return err
 	} else if err != nil {
-		return errors.Wrap(err, "can't add message")
+		return errors.Wrap(err, "can't add a message")
 	}
 	return nil
 }
 
-type AsyncStream interface {
-	Start() error
-	Stop()
-}
-
+// Stream converts messages subscription model to a stream of messages.
 type Stream struct {
-	options protocol.SubscribeOptions
+	sync.Mutex
```

Comment:
can you put it on top of cancel? it guards it right? i would also hide it to prevent access

Reply:
+1 to hide it.

> can you put it on top of cancel? it guards it right?

Yes.